Add ERC: Cache invalidation in ERC-5219 mode Web3 URL (#652)
ethereum/ERCs: ERCS/erc-7774.md

---
eip: 7774
title: Cache invalidation in ERC-5219 mode Web3 URL

> description: In ERC-5219 resolve mode, add mechanisms to alleviate limitations that prevent the use of RFC 9111 HTTP caching

Reviewer: Your description is a bit flowery. How about:

```suggestion
description: Fix ERC-5219's resolve mode to allow use of RFC 9111 HTTP caching
```

Reviewer: After reading this, how about:

```suggestion
description: Introduce a new HTTP cache control method using EVM events for invalidation
```

However, due to the [ERC-6944](./eip-6944.md) resolve mode not forwarding request HTTP headers to the smart contract, smart contracts cannot handle `If-None-Match` and `If-Modified-Since` cache validation headers. Consequently, they are limited to using the `Cache-control: max-age=XX` mechanism, which causes each cache validation request to trigger an RPC call, regenerating the full response.

This ERC proposes a solution to bypass this limitation by allowing websites to broadcast cache invalidations via smart contract events.

Additionally, even if smart contracts could read request HTTP headers, using smart contract events is more efficient, as it moves cache invalidation logic outside the contract.

## Specification

This standard introduces the `evm-events` cache directive for the `Cache-Control` header of request responses, as an extension directive as defined in section 5.2.3 of [RFC 9111](https://www.rfc-editor.org/rfc/rfc9111).

> When a [ERC-6944](./eip-6944.md) resolve mode website wants to use event-based caching for a request, it MUST:

Reviewer: ```suggestion
When an [ERC-6944](./eip-6944.md) resolve mode website wants to use event-based caching for a request, it MUST:
```

     - If the `ETag` in the mapping matches the `If-None-Match` value, the `web3://` client returns a `304 Not Modified` response immediately.
     - If the `ETag` does not match, the client queries the smart contract, deletes the mapping entry, and processes the request as if no mapping entry existed.
   
   - If the request contains a valid `If-Modified-Since` header:
     - If the `last modified date` in the mapping is earlier than the `If-Modified-Since` date, the client returns a `304 Not Modified` response immediately.
     - Otherwise, the client queries the smart contract, deletes the mapping entry, and processes the request as if no mapping entry existed.
   
   - If the request contains neither `If-None-Match` nor `If-Modified-Since` headers (or they are invalid):
     - The client queries the smart contract, deletes the mapping entry, and processes the request as if no mapping entry existed.

#### Cache Invalidation via Blockchain Events

In the "Listening for Events" state, the `web3://` client listens to the blockchain for the cache invalidation events defined in the previous section. For each path match, it deletes the corresponding mapping entry.

## Rationale

> We add this feature to the [ERC-6944](./eip-6944.md) resolve mode because it can be added without changes the interface.

Reviewer: This line is more motivation than rationale.

Author: Thanks! The line between motivation and rationale is always a bit blurry :)

To stay as close as possible to standard HTTP, we reuse the HTTP caching mechanism headers.

> The use of the `evm-events` directive is necessary to avoid a situation where a website uses traditional [RFC 9111](https://www.rfc-editor.org/rfc/rfc9111) HTTP caching headers, but does not implement this ERC  by failing to emit the events. In such cases, `web3://` clients implementing this ERC would serve stale content for that website indefinitely.

Reviewer: ```suggestion
The use of the `evm-events` directive is necessary to avoid a situation where a website uses traditional [RFC 9111](https://www.rfc-editor.org/rfc/rfc9111) HTTP caching headers, but the contract does not implement this ERC  by failing to emit the events. In such cases, `web3://` clients implementing this ERC would serve stale content for that website indefinitely.
```

Is this more clear and still correct?

## Security Considerations

> No security considerations were found.

Reviewer: There are definitely some security considerations, no? At the very least contracts need to correctly emit cache invalidation events correctly, or stale content could be served.

I bet there are some unpleasant interactions between this and chain reorgs. as well?

Author: Yes indeed! 
I have added 3 : 

- Stale content will be served during the delay between a user transaction emitting a cache clearing event, and the `web3://` client picking and processing the event.
- For each cached page, websites must properly implement cache invalidation events; otherwise, stale content will be served indefinitely.
- In the event of a chain reorganization, the `web3://` client must roll back its caching state, or reverted content will be served until the next cache clearing event.

The first one is something that conceptually annoys me (even if the `web3://` client is optimizing its logs reading by using streams, it can theorically be missed).
So I am thinking about ways for the website to signal to the `web3://` client when fetching a page: "hey please ensure you are up to date with block XXX (because I just made a change on it via a TX)".
A HTTP header in the request is an idea, but may be not practical when doing a page change (e.g. in blog post edit page, save tx, then go to blog post page view). Thinking...
